src/docfx/lib/process/InterProcessReaderWriterLock.cs

```diff
@@ -0,0 +1,108 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Diagnostics;
+using System.IO;
+using System.Threading;
+
+namespace Microsoft.Docs.Build
+{
+    public struct InterProcessReaderWriterLock : IDisposable
+    {
+        private readonly FileStream _fileStream;
+        private readonly string _filePath;
+
+        private InterProcessReaderWriterLock(FileStream fileStream, string filePath)
+        {
+            _fileStream = fileStream;
+            _filePath = filePath;
+        }
+
+        public static InterProcessReaderWriterLock CreateReaderLock(string lockName)
+        {
+            Debug.Assert(!string.IsNullOrEmpty(lockName));
+
+            var filePath = Path.Combine(AppData.MutexRoot, HashUtility.GetMd5Hash(lockName));
+            var fileLock = WaitFile(lockName, filePath, FileAccess.Read, FileShare.Read);
+
+            return new InterProcessReaderWriterLock(fileLock, filePath);
+        }
+
+        public static InterProcessReaderWriterLock CreateWriterLock(string lockName)
+        {
+            Debug.Assert(!string.IsNullOrEmpty(lockName));
+
+            var filePath = Path.Combine(AppData.MutexRoot, HashUtility.GetMd5Hash(lockName));
+            var fileLock = WaitFile(lockName, filePath, FileAccess.Write, FileShare.None);
+
+            return new InterProcessReaderWriterLock(fileLock, filePath);
+        }
+
+        private static FileStream WaitFile(string name, string path, FileAccess access, FileShare fileShare)
+        {
+            var start = DateTime.UtcNow;
+            while (DateTime.UtcNow - start <= TimeSpan.FromMinutes(1))
```

Comment:
suggest to repeat forever until we grab the lock, similar to `InterProcessMutex`